vcardtools.py

```diff
@@ -111,6 +116,25 @@ def init_parser():
     )
     return parser
 
+def sanitise_name(a_name: str) -> str:
+    """ Sanitise the name, basically a filename,
+        by removing characters which would cause a problem when creating a file in the OS
+        and replacing them with something safe (in this case, an underscore)
+    """
+    NEW_REPLACEMENT_CHAR = '_'
+    FROM_CHARACTERS = ' .\\/"\'!@#?$%^&*|()[]{};:<>'
+    for old_char in FROM_CHARACTERS:
+        a_name = a_name.replace(old_char, NEW_REPLACEMENT_CHAR)
```

Comment:
Hey @tonysutherland,
On that one my opinion is that thoses options belong only to the *CLI*, not the *lib* so I would definitively put them in the *CLI* file (i.e. *vcardtools.py*) not the *lib* one.

Side note: I think I should refactor all those options into a configuration *dict* in the library and allow the *CLI* to overwrite the default values. I don't like having the *__main__* script setting the module's variables.
What do you think about that ?
I just don't have enough time these days... :sweat: I have already planned to improve its quality by running more linters and formatters on it.

Reply:
So, I have written a little utility function(s) which basically does what you want to do in relation to a configuration dict.
Basically, it creates a configuration dict which contains all the values used in the program as defaults or constants etc. Simple key/value pair with key=name of variable and value=value of variable.
It gets them from these sources, in this order:
1. a dict of initial default values (set somewhere in the program, usually at the start)
2. a configuration file, in YAML format (which is a superset of JSON, so it can be in JSON format).  Actually, the routine accepts a list of configuration files, so that multiple different configurations can be specified (eg a standard config file, say in ~/.config and then an over-riding config file say specified on the command line)
3. all the current environment variables (this can be specified as OFF, so that the configuration is not polluted with lots of environment variables, but it is not a problem using it, since mostly you just ignore what is not needed and just use the environment variables which you need
4. the command line arguments.

The order means that the later items over-ride the earlier ones - which is basically what one would want normally.
You can always set the configuration dict item later (after loading all the above) in whatever manner you feel like. There is a little of this in the code you have in main()

The routine also does variable substitution.
Variables are entered as ${varname} (similar to how Make works), which is replaced by the definition of the variable.
The definition of the variable is an entry in the dictionary.

Once there is a configuration dict, then I simply pass that variable (I usually call it "config" or "ctx" (stands for context) into routines that need to use it (because I don't like having global variables used in functions), but this variable could easily be specified as a global variable and used in any routine.

If you want it to be used in the library, it would be best to define it as a global in the library and set that global from the configuration function.

I will provide the code as a library, and update your code to use it and comment out the original code, so you can see what it does.

Might take me a bit of time to do all this.

Reply:
I would rather add an option to inject custom regex from CLI, even with the possibility to use some predefined regex.

Also, for more Windows-compatibility maybe this regex can be used: `\A(?!(?:COM[0-9]|CON|LPT[0-9]|NUL|PRN|AUX|com[0-9]|con|lpt[0-9]|nul|prn|aux)|\s|[\.]{2,})[^\\\/:*"?<>|]{1,254}(?<![\s\.])\z`

[Source for regex](https://stackoverflow.com/a/45615798)

Reply:
@RaistGH Yes injecting a custom regex is a good idea :+1: 
Through I would still maintain the predefined regex with the option to use underscore instead of space for a better UX.
Thank you very much for the Windows compatible regex ! :pray:

Reply:
I have implemented everything we discussed here in ac918a12dfa87a32683abbfdf6047341bb54eea0 d686dd7afb30b4ff317b013ef5acbb49c4dd8bd9 0ac32c628f48ed99836ee90585eaf266a42cbf06 but used the _windows valid filename regex_ mentionned by @RaistGH nor introduced a _user CLI option for custom regex_.
I might do this later (way later), or if you have a way to make another PR with those changes I would be happy to merge it :wink: